File: BankersAlgorithm.java
import java.util.Scanner;

public class BankersAlgorithm {
    // Variables for number of processes and resources
    private int numProcesses;
    private int numResources;

    // Matrices and arrays for the Banker's algorithm
    private int[][] allocation; // Allocation matrix: resources currently allocated to each process
    private int[][] max;        // Max matrix: maximum demand of each process
    private int[][] need;       // Need matrix: additional resources each process needs to complete
    private int[] available;    // Available array: resources currently available in the system

    // Constructor to initialize matrices and arrays
    public BankersAlgorithm(int processes, int resources) {
        numProcesses = processes; // Number of processes
        numResources = resources; // Number of resources
        allocation = new int[processes][resources]; // Allocation matrix initialization
        max = new int[processes][resources];        // Max matrix initialization
        need = new int[processes][resources];       // Need matrix initialization
        available = new int[resources];            // Available array initialization
    }

    // Method to input allocation, max, and available matrices from the user
    public void inputMatrices() {
        Scanner scanner = new Scanner(System.in);

        // Input Allocation matrix
        System.out.println("Enter Allocation matrix: ");
        for (int i = 0; i < numProcesses; i++) {
            for (int j = 0; j < numResources; j++) {
                allocation[i][j] = scanner.nextInt();
            }
        }

        // Input Max matrix
        System.out.println("Enter Max matrix: ");
        for (int i = 0; i < numProcesses; i++) {
            for (int j = 0; j < numResources; j++) {
                max[i][j] = scanner.nextInt();
            }
        }

        // Input Available resources array
        System.out.println("Enter Available resources: ");
        for (int i = 0; i < numResources; i++) {
            available[i] = scanner.nextInt();
        }

        // Calculate the Need matrix: `Need = Max - Allocation`
        for (int i = 0; i < numProcesses; i++) {
            for (int j = 0; j < numResources; j++) {
                need[i][j] = max[i][j] - allocation[i][j];
            }
        }

        scanner.close();
    }

    // Method to check if the system is in a safe state
    public boolean checkSafeState() {
        boolean[] finished = new boolean[numProcesses]; // Tracks if a process is finished
        int[] work = new int[numResources]; // Work array to simulate available resources
        System.arraycopy(available, 0, work, 0, numResources); // Copy available to work array

        StringBuilder safeSequence = new StringBuilder(); // To store the safe sequence
        int count = 0; // Number of finished processes

        // Loop until all processes are checked or no further allocation is possible
        while (count < numProcesses) {
            boolean found = false;

            // Check each process to see if its need can be satisfied
            for (int i = 0; i < numProcesses; i++) {
                if (!finished[i]) { // Process is not yet finished
                    boolean canAllocate = true;

                    // Check if the current process's need can be met
                    for (int j = 0; j < numResources; j++) {
                        if (need[i][j] > work[j]) {
                            canAllocate = false; // Need exceeds available resources
                            break;
                        }
                    }

                    if (canAllocate) {
                        // Allocate resources to process i and mark it as finished
                        for (int j = 0; j < numResources; j++) {
                            work[j] += allocation[i][j]; // Release resources after completion
                        }
                        safeSequence.append(i).append(" "); // Add process to safe sequence
                        finished[i] = true;
                        found = true;
                        count++;
                    }
                }
            }

            if (!found) {
                // If no process could be allocated resources, the system is unsafe
                System.out.println("System is not in a safe state.");
                return false;
            }
        }

        // If all processes are finished, the system is in a safe state
        System.out.println("System is in a safe state. Safe sequence is: " + safeSequence);
        return true;
    }

    // Main method to execute the Banker's Algorithm
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Input number of processes and resources
        System.out.println("Enter the number of processes: ");
        int processes = scanner.nextInt();

        System.out.println("Enter the number of resources: ");
        int resources = scanner.nextInt();

        // Create a Banker's Algorithm object and input matrices
        BankersAlgorithm bankers = new BankersAlgorithm(processes, resources);
        bankers.inputMatrices();

        // Check and display if the system is in a safe state
        bankers.checkSafeState();

        scanner.close();
    }
}


// Commands to run:
// javac BankersAlgorithm.java
// java BankersAlgorithm

// Sample input:
/*
Enter the number of processes: 
5
Enter the number of resources: 
3
Enter Allocation matrix: 
0 1 0
2 0 0
3 0 2
2 1 1
0 0 2
Enter Max matrix: 
7 5 3
3 2 2
9 0 2
2 2 2
4 3 3
Enter Available resources: 
3 3 2
*/

File: BankersAlgorithm.c
// Bankers algo
#include <stdio.h>
#include <stdbool.h>

#define MAX_PROCESSES 5
#define MAX_RESOURCES 3

// Function to check if the system is in a safe state
bool isSafeState(int processes[], int avail[], int max[][MAX_RESOURCES], int alloc[][MAX_RESOURCES], int need[][MAX_RESOURCES], int num_processes, int num_resources) {
    bool finish[MAX_PROCESSES] = {false};
    int safeSequence[MAX_PROCESSES];
    int work[MAX_RESOURCES];
    
    // Initialize work as available resources
    for (int i = 0; i < num_resources; i++) {
        work[i] = avail[i];
    }

    int count = 0;
    
    // Find the processes that can finish
    while (count < num_processes) {
        bool found = false;
        for (int p = 0; p < num_processes; p++) {
            if (!finish[p]) {
                int j;
                for (j = 0; j < num_resources; j++) {
                    if (need[p][j] > work[j])
                        break;
                }
                
                // If all resources for process p can be allocated
                if (j == num_resources) {
                    // Process p can finish
                    for (int k = 0; k < num_resources; k++) {
                        work[k] += alloc[p][k];
                    }
                    safeSequence[count++] = p;
                    finish[p] = true;
                    found = true;
                }
            }
        }

        // If no process could be found in this iteration, system is not in a safe state
        if (!found) {
            return false;
        }
    }
    
    // If all processes can finish, system is in a safe state
    printf("System is in a safe state.\nSafe Sequence: ");
    for (int i = 0; i < num_processes; i++) {
        printf("%d ", safeSequence[i]);
    }
    printf("\n");
    return true;
}

// Function to calculate the need matrix
void calculateNeed(int need[][MAX_RESOURCES], int max[][MAX_RESOURCES], int alloc[][MAX_RESOURCES], int num_processes, int num_resources) {
    for (int i = 0; i < num_processes; i++) {
        for (int j = 0; j < num_resources; j++) {
            need[i][j] = max[i][j] - alloc[i][j];
        }
    }
}

int main() {
    int num_processes, num_resources;
    
    printf("Enter the number of processes: ");
    scanf("%d", &num_processes);

    printf("Enter the number of resources: ");
    scanf("%d", &num_resources);

    int processes[MAX_PROCESSES], avail[MAX_RESOURCES], max[MAX_PROCESSES][MAX_RESOURCES], alloc[MAX_PROCESSES][MAX_RESOURCES], need[MAX_PROCESSES][MAX_RESOURCES];

    // Input the available resources
    printf("Enter the available resources: \n");
    for (int i = 0; i < num_resources; i++) {
        scanf("%d", &avail[i]);
    }

    // Input the maximum demand of each process
    printf("Enter the maximum demand of each process: \n");
    for (int i = 0; i < num_processes; i++) {
        for (int j = 0; j < num_resources; j++) {
            scanf("%d", &max[i][j]);
        }
    }

    // Input the allocated resources for each process
    printf("Enter the allocated resources for each process: \n");
    for (int i = 0; i < num_processes; i++) {
        for (int j = 0; j < num_resources; j++) {
            scanf("%d", &alloc[i][j]);
        }
    }

    // Calculate the need matrix
    calculateNeed(need, max, alloc, num_processes, num_resources);

    // Check if the system is in a safe state
    if (!isSafeState(processes, avail, max, alloc, need, num_processes, num_resources)) {
        printf("System is not in a safe state.\n");
    }

    return 0;
}


// Sample Input:
/*
Enter the number of processes: 5
Enter the number of resources: 3
Enter the available resources: 
3 3 2
Enter the maximum demand of each process: 
7 5 3
3 2 2
9 0 2
2 2 2
4 3 3 
Enter the allocated resources for each process: 
0 1 0
2 0 0
3 0 2
2 1 1
0 0 2
*/